File: dlist.go
package main

import (
	"fmt"
)

type DNode struct {
	data string
	next *DNode
	prev *DNode
}

type DList struct {
	head *DNode
	tail *DNode
	size int
}

func NewDList() *DList {
	return &DList{
		head: nil,
		tail: nil,
		size: 0,
	}
}

func (l *DList) DLCLEAR() {
	l.head = nil
	l.tail = nil
	l.size = 0
	fmt.Println("Двусвязный список очищен.")
}

func (l *DList) DLPUSH_HEAD(value string) {
	newNode := &DNode{
		data: value,
		next: l.head,
		prev: nil,
	}

	if l.head != nil {
		l.head.prev = newNode
	} else {
		l.tail = newNode
	}

	l.head = newNode
	l.size++
}

func (l *DList) DLPUSH_TAIL(value string) {
	newNode := &DNode{
		data: value,
		next: nil,
		prev: l.tail,
	}

	if l.tail != nil {
		l.tail.next = newNode
	} else {
		l.head = newNode
	}

	l.tail = newNode
	l.size++
}

func (l *DList) DLPUSH_BEFORE(target, value string) {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	current := l.head
	for current != nil && current.data != target {
		current = current.next
	}

	if current == nil {
		return
	}

	if current == l.head {
		l.DLPUSH_HEAD(value)
		return
	}

	newNode := &DNode{
		data: value,
		next: current,
		prev: current.prev,
	}
	current.prev.next = newNode
	current.prev = newNode
	l.size++
}

func (l *DList) DLPUSH_AFTER(target, value string) {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	current := l.head
	for current != nil && current.data != target {
		current = current.next
	}

	if current == nil {
		return
	}

	if current == l.tail {
		l.DLPUSH_TAIL(value)
		return
	}

	newNode := &DNode{
		data: value,
		next: current.next,
		prev: current,
	}
	current.next.prev = newNode
	current.next = newNode
	l.size++
}

func (l *DList) DLREMOVE_HEAD() {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	if l.head == l.tail {
		l.head = nil
		l.tail = nil
	} else {
		l.head = l.head.next
		l.head.prev = nil
	}
	l.size--
}

func (l *DList) DLREMOVE_TAIL() {
	if l.tail == nil {
		fmt.Println("Список пуст!")
		return
	}

	if l.head == l.tail {
		l.head = nil
		l.tail = nil
	} else {
		l.tail = l.tail.prev
		l.tail.next = nil
	}
	l.size--
}

func (l *DList) DLREMOVE_BEFORE(target string) {
	if l.head == nil || l.head.next == nil {
		fmt.Println("Недостаточно элементов для удаления!")
		return
	}

	current := l.head
	for current != nil && current.data != target {
		current = current.next
	}

	if current == nil || current.prev == nil {
		return
	}

	if current.prev == l.head {
		l.DLREMOVE_HEAD()
		return
	}

	nodeToRemove := current.prev
	nodeToRemove.prev.next = current
	current.prev = nodeToRemove.prev
	l.size--
}

func (l *DList) DLREMOVE_AFTER(target string) {
	if l.head == nil || l.head.next == nil {
		fmt.Println("Недостаточно элементов для удаления!")
		return
	}

	current := l.head
	for current != nil && current.data != target {
		current = current.next
	}

	if current == nil || current.next == nil {
		return
	}

	if current.next == l.tail {
		l.DLREMOVE_TAIL()
		return
	}

	nodeToRemove := current.next
	current.next = nodeToRemove.next
	nodeToRemove.next.prev = current
	l.size--
}

func (l *DList) DLREMOVE_VALUE(value string) {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	current := l.head
	for current != nil && current.data != value {
		current = current.next
	}

	if current == nil {
		return
	}

	if current == l.head {
		l.DLREMOVE_HEAD()
	} else if current == l.tail {
		l.DLREMOVE_TAIL()
	} else {
		current.prev.next = current.next
		current.next.prev = current.prev
		l.size--
	}
}

func (l *DList) DLSEARCH(value string) bool {
	current := l.head
	for current != nil {
		if current.data == value {
			return true
		}
		current = current.next
	}
	return false
}

func (l *DList) DLGET(index int) (string, error) {
	if index >= l.size || index < 0 {
		return "", fmt.Errorf("индекс %d вне диапазона [0, %d]", index, l.size-1)
	}

	var current *DNode
	// Оптимизация: идем с начала или с конца в зависимости от индекса
	if index < l.size/2 {
		current = l.head
		for i := 0; i < index; i++ {
			current = current.next
		}
	} else {
		current = l.tail
		for i := l.size - 1; i > index; i-- {
			current = current.prev
		}
	}
	return current.data, nil
}

func (l *DList) DLPRINT_FORWARD() {
	if l.head == nil {
		fmt.Println("Список пуст!")
		return
	}

	fmt.Printf("Двусвязный список [%d] (прямой порядок): NULL <- ", l.size)
	current := l.head
	for current != nil {
		fmt.Printf("\"%s\"", current.data)
		if current.next != nil {
			fmt.Print(" <-> ")
		}
		current = current.next
	}
	fmt.Println(" -> NULL")
}

func (l *DList) DLPRINT_BACKWARD() {
	if l.tail == nil {
		fmt.Println("Список пуст!")
		return
	}

	fmt.Printf("Двусвязный список [%d] (обратный порядок): NULL <- ", l.size)
	current := l.tail
	for current != nil {
		fmt.Printf("\"%s\"", current.data)
		if current.prev != nil {
			fmt.Print(" <-> ")
		}
		current = current.prev
	}
	fmt.Println(" -> NULL")
}

func (l *DList) DLLENGTH() int {
	return l.size
}

// Для сериализации
func (l *DList) GetAll() []string {
	result := make([]string, 0, l.size)
	current := l.head
	for current != nil {
		result = append(result, current.data)
		current = current.next
	}
	return result
}

func (l *DList) SetAll(items []string) {
	l.DLCLEAR()
	for _, item := range items {
		l.DLPUSH_TAIL(item)
	}
}

File: DList.cpp
#include "DList.h"
#include <iostream>
#include <fstream>
#include <vector>
#include <stdexcept>

using namespace std;

DList::DList() {
    head = nullptr;
    tail = nullptr;
    size = 0;
}

DList::~DList() {
    DLCLEAR();
}

void DList::DLCLEAR() {
    DNode* current = head;
    while (current != nullptr) {
        DNode* temp = current;
        current = current->next;
        delete temp;
    }
    head = nullptr;
    tail = nullptr;
    size = 0;
    cout << "Двусвязный список очищен." << endl;
}

void DList::DLPUSH_HEAD(const string& value) {
    DNode* newNode = new DNode{ value, head, nullptr };

    if (head != nullptr) {
        head->prev = newNode;
    }
    else {
        tail = newNode;
    }

    head = newNode;
    size++;
    //cout << "Элемент \"" << value << "\" добавлен в голову списка" << endl;
}

void DList::DLPUSH_TAIL(const string& value) {
    DNode* newNode = new DNode{ value, nullptr, tail };

    if (tail != nullptr) {
        tail->next = newNode;
    }
    else {
        head = newNode;
    }

    tail = newNode;
    size++;
    //cout << "Элемент \"" << value << "\" добавлен в хвост списка" << endl;
}

void DList::DLPUSH_BEFORE(const string& target, const string& value) {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    DNode* current = head;
    while (current != nullptr && current->data != target) {
        current = current->next;
    }

    if (current == nullptr) {
        //cout << "Элемент \"" << target << "\" не найден в списке" << endl;
        return;
    }

    if (current == head) {
        DLPUSH_HEAD(value);
        return;
    }

    DNode* newNode = new DNode{ value, current, current->prev };
    current->prev->next = newNode;
    current->prev = newNode;
    size++;
    //cout << "Элемент \"" << value << "\" добавлен перед \"" << target << "\"" << endl;
}

void DList::DLPUSH_AFTER(const string& target, const string& value) {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    DNode* current = head;
    while (current != nullptr && current->data != target) {
        current = current->next;
    }

    if (current == nullptr) {
        //cout << "Элемент \"" << target << "\" не найден в списке" << endl;
        return;
    }

    if (current == tail) {
        DLPUSH_TAIL(value);
        return;
    }

    DNode* newNode = new DNode{ value, current->next, current };
    current->next->prev = newNode;
    current->next = newNode;
    size++;
    //cout << "Элемент \"" << value << "\" добавлен после \"" << target << "\"" << endl;
}

void DList::DLREMOVE_HEAD() {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    DNode* temp = head;
    head = head->next;

    if (head != nullptr) {
        head->prev = nullptr;
    }
    else {
        tail = nullptr;
    }

    //cout << "Элемент \"" << temp->data << "\" удалён из головы списка" << endl;
    delete temp;
    size--;
}

void DList::DLREMOVE_TAIL() {
    if (tail == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    DNode* temp = tail;
    tail = tail->prev;

    if (tail != nullptr) {
        tail->next = nullptr;
    }
    else {
        head = nullptr;
    }

    //cout << "Элемент \"" << temp->data << "\" удалён из хвоста списка" << endl;
    delete temp;
    size--;
}

void DList::DLREMOVE_BEFORE(const string& target) {
    if (head == nullptr || head->next == nullptr) {
        cout << "Недостаточно элементов для удаления!" << endl;
        return;
    }

    DNode* current = head;
    while (current != nullptr && current->data != target) {
        current = current->next;
    }

    if (current == nullptr) {
        //cout << "Элемент \"" << target << "\" не найден в списке" << endl;
        return;
    }

    if (current->prev == nullptr) {
        cout << "Невозможно удалить элемент перед головой!" << endl;
        return;
    }

    if (current->prev == head) {
        DLREMOVE_HEAD();
        return;
    }

    DNode* temp = current->prev;
    temp->prev->next = current;
    current->prev = temp->prev;

    //cout << "Элемент \"" << temp->data << "\" удалён перед \"" << target << "\"" << endl;
    delete temp;
    size--;
}

void DList::DLREMOVE_AFTER(const string& target) {
    if (head == nullptr || head->next == nullptr) {
        cout << "Недостаточно элементов для удаления!" << endl;
        return;
    }

    DNode* current = head;
    while (current != nullptr && current->data != target) {
        current = current->next;
    }

    if (current == nullptr) {
        //cout << "Элемент \"" << target << "\" не найден в списке" << endl;
        return;
    }

    if (current->next == nullptr) {
        cout << "Невозможно удалить элемент после хвоста!" << endl;
        return;
    }

    if (current->next == tail) {
        DLREMOVE_TAIL();
        return;
    }

    DNode* temp = current->next;
    current->next = temp->next;
    temp->next->prev = current;

    //cout << "Элемент \"" << temp->data << "\" удалён после \"" << target << "\"" << endl;
    delete temp;
    size--;
}

void DList::DLREMOVE_VALUE(const string& value) {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }
    DNode* current = head;

    while (current != nullptr && current->data != value) {
        current = current->next;
    }

    if (current == nullptr) {
        //cout << "Элемент \"" << value << "\" не найден в списке" << endl;
        return;
    }

    if (current == head) {
        head = current->next;
        if (head != nullptr) {
            head->prev = nullptr;
        }
        else {
            tail = nullptr;
        }

        //cout << "Элемент \"" << current->data << "\" (голова) удалён по значению" << endl;
        delete current;
        size--;
        return;
    }

    if (current == tail) {
        tail = current->prev;
        if (tail != nullptr) {
            tail->next = nullptr;
        }
        else {
            head = nullptr;
        }

        //cout << "Элемент \"" << current->data << "\" (хвост) удалён по значению" << endl;
        delete current;
        size--;
        return;
    }

    current->prev->next = current->next;
    current->next->prev = current->prev;

    //cout << "Элемент \"" << current->data << "\" удалён по значению" << endl;
    delete current;
    size--;
}

bool DList::DLSEARCH(const string& value) const {
    DNode* current = head;
    size_t index = 0;
    while (current != nullptr) {
        if (current->data == value) {
            //cout << "Элемент \"" << value << "\" найден на позиции " << index << endl;
            return true;
        }
        current = current->next;
        index++;
    }
    //cout << "Элемент \"" << value << "\" не найден в списке" << endl;
    return false;
}

string DList::DLGET(size_t index) const {
    if (index >= size) {
        throw out_of_range("Индекс " + to_string(index) + " вне диапазона [0, " + to_string(size - 1) + "]");
    }

    DNode* current = head;
    for (size_t i = 0; i < index; i++) {
        current = current->next;
    }
    return current->data;
}

void DList::DLPRINT_FORWARD() const {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    cout << "Двусвязный список [" << size << "] (прямой порядок): NULL <- ";
    DNode* current = head;
    while (current != nullptr) {
        cout << "\"" << current->data << "\"";
        if (current->next != nullptr) {
            cout << " <-> ";
        }
        current = current->next;
    }
    cout << " -> NULL" << endl;
}

void DList::DLPRINT_BACKWARD() const {
    if (tail == nullptr) {
        cout << "Список пуст!" << endl;
        return;
    }

    cout << "Двусвязный список [" << size << "] (обратный порядок): NULL <- ";
    DNode* current = tail;
    while (current != nullptr) {
        cout << "\"" << current->data << "\"";
        if (current->prev != nullptr) {
            cout << " <-> ";
        }
        current = current->prev;
    }
    cout << " -> NULL" << endl;
}

void DList::DLPRINT_HEAD() const {
    if (head == nullptr) {
        cout << "Список пуст!" << endl;
    }
    else {
        cout << "Голова списка: \"" << head->data << "\"" << endl;
    }
}

void DList::DLPRINT_TAIL() const {
    if (tail == nullptr) {
        cout << "Список пуст!" << endl;
    }
    else {
        cout << "Хвост списка: \"" << tail->data << "\"" << endl;
    }
}

size_t DList::DLLENGTH() const {
    return size;
}
